sympy/sets/handlers/issubset.py

```diff
@@ -66,6 +66,18 @@ def is_subset_sets(a_range, b_interval): # noqa:F811
             cond_right = a_range.sup <= b_interval.right
         return fuzzy_and([cond_left, cond_right])
 
+@dispatch(Range, FiniteSet)
+def is_subset_sets(a_range, b_finiteset): # noqa:F811
+    try:
+        a_size = a_range.size
+    except ValueError:
+        # symbolic Range of unknown size
+        return None
+    if a_size > len(b_finiteset):
+        return False
+    else:
+        return fuzzy_and(b_finiteset.contains(x) for x in a_range)
```

Comment:
I guess we could use this logic for any finite iterable set rather than just `Range`. It could be slow though e.g.:
```
In [5]: S = FiniteSet(*(Dummy() for _ in range(10**3)))

In [6]: Range(10**3).is_subset(S) # Really slow...
```

Reply:
May be this can be analysed just by using endpoints of `FiniteSet` and `Range`. @gschintgen Have you given this idea a consideration?

Reply:
> Range(10**3).is_subset(S) # Really slow...

That's a pity. Even a *single* `S.contains(5)` takes almost 2 seconds (on a slow system). I concede that I must have imagined `FiniteSet` to be hash-based so that a contains-check would be almost instant (and constant-time relative to the cardinality). But in hindsight it's obvious that, in order to give mathematically meaningful answers, `contains` must indeed check equality with every element of the `FiniteSet`. So it comes down to 2ms per equality check which sums up rather quickly. Hmm. (2ms is still expensive for a rather basic check, isn't it?)
So the problem with the proposed code is that its running time scales with the product of the cardinalities of both sets and that's not really acceptable.

@czgdp1807 : I'm not sure I completely understand your idea. A FiniteSet may contain an arbitrary number of arbitrary expressions. What could potentially be done is somehow filtering/pre-processing the FiniteSet in a single run-through (e.g. kick out irrelevant non-integers). I'll think about it.

Reply:
Oh well, this took me some time. I've added a new commit that changes the logic. I just hope I got my testing right. Please review, but note the gotcha below that I discovered only very late...

With this new commit, the example given by @oscarbenjamin is handled almost instantly by the handler, **but** there's another **huge** slowdown in line 385:
https://github.com/sympy/sympy/blob/f7390e5fa2313be1c49959270358b2b0c62a1a57/sympy/sets/sets.py#L377-L386
The `None` case is already pathologically slow in current master due to the intersection computation.

I also note that the logic of that intersection test is bound to fail for Range.intersect(FiniteSet): The result will be a FiniteSet and hence test unequal to self. So the expensive computation is in vain.

So to sum up (for the slow example):
master: slow due to useless intersection computation
first commit: slow handler, followed by slow intersection
new commit: faster handler, followed by slow intersection.

Reply:
Not sure if I added that comment but it sounds like something I'd say...

Reply:
I wonder if it's possible to speed up general membership testing with something like this. Maybe the `FiniteSet` could cache a `set` that can be used for fast membership testing as separate from the more complicated symbolic cases.

Reply:
I've added a basic cache. (In fact a temporary `set` of all elements/args was created anyway in `__new__`, so it was sufficient to keep it around.) For larger `FiniteSets` it's indeed beneficial to first check membership using this cache. But only the `True` case with an exact match is sped up.
(See commit message for a cursory benchmark.)

Reply:
It doesn't need to be done in this PR but I guess there are some situations where we know based on type that `==` is sufficient and therefore can use efficient set algorithms. For example if all elements of both sets are of type `Number` then we can process everything using `set` because we know that `Eq(a, b)` is True if `a==b` and False otherwise (not sure about `Float`...). On other hand if any element is a symbol with no defined assumptions then we know that `Eq(a, b)` will be True if `a == b` and "None" otherwise so probably something can be done with that as well.

Reply:
I think we've probably over-complicated things here. Presumably the simple contains loop would cover the case
```
Range(n, n + 2).is_subset(FiniteSet(n, n + 1))
```
which with the PR now is None rather than True.

I guess we should at least fall back on the slow method here. It's probably more important to do the right thing with small symbolic sets than worry about performance with larger sets.

Reply:
> ```
> Range(n, n + 2).is_subset(FiniteSet(n, n + 1))
> ```
> 
> 
> which with the PR now is None rather than True.

I agree that for a CAS symbolics should be more important than pure performance. I'm somewhat astonished that in the early commit with the basic `fuzzy_and` check I included this test with `None` as expected result. (The code of the very first commit indeed returns `True`!) I know I did some trials with symbolic `Ranges` but must have misinterpreted something. Possibly a mishap with the `n` in the tests file that doesn't have integer assumptions combined with #18146 cropping up. I think I even concluded that `Range` and symbols don't mix well...

I'll see what I can do.

Reply:
@oscarbenjamin, @czgdp1807 : What do you think now with the latest commits?

As for the more advanced caching, it's an interesting idea and I'm sure there's more that could be done (e.g. not only check if all elements are of certain types, but maybe cache precisely which `args` are of other/symbolic types, so that only those specific elements need to be checked.) But at this stage I'm not sure how important this optimization for large `FiniteSets` is. I'd rather weed out logic errors, sets-related or not.

Reply:
> I'd rather weed out logic errors, sets-related or not.

Agreed